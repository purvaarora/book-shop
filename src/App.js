/*
To try and encourage more sales of different books from a popular 5 book series, a bookshop has decided to offer discounts on multiple book 
purchases. One copy of any of the five books costs $8. If, however, you buy two different books, you get a 5% discount on those two books. 
If you buy 3 different books, you get a 10% discount. If you buy 4 different books, you get a 20% discount. 
If you buy all 5, you get a 25% discount.

Note: that if you buy four books, of which 3 are different titles, you get a 10% discount on the 3 that form part of a set, but the fourth book 
still costs $8. Your mission is to write a piece of code to calculate the price of any conceivable shopping basket (containing only books of 
the same series), giving as big a discount as possible.
For example, how much does this basket of books cost?
2 copies of the first book 2 copies of the second book 2 copies of the third book 1 copy of the fourth book 1 copy of the fifth book
*/

import React from 'react'
import map from 'lodash/map'
import find from 'lodash/find'
import get from 'lodash/get'
import cloneDeep from 'lodash/cloneDeep'
import values from 'lodash/values'
import sum from 'lodash/sum'
import PropTypes from 'prop-types'

import ListItem from './components/ListItem'

class BookStore extends React.Component {
  constructor(props) {
    super(props);
    this.baseFare = 8; // Base fare for each book (All books have same fare for this case)
    this.state = {
      shoppingCart: {}, // to maintain books added by guest
      amount: 0, // final amount of cart
    }
    this.handleAddToCart = this.handleAddToCart.bind(this)
    this.handleRemoveFromCart = this.handleRemoveFromCart.bind(this)
    this.handleTotalBill = this.handleTotalBill.bind(this)
  }

  // Method to add a book to shopping cart
  handleAddToCart(id) {
    let shoppingCart = this.state.shoppingCart
    if (!shoppingCart[id]) {
      shoppingCart[id] = 1
    } else {
      shoppingCart[id]++
    }
    this.setState({
      shoppingCart
    }, () => {
      this.handleTotalBill()
    })
  }

  // Method to remove a book from shopping cart
  handleRemoveFromCart(id) {
    let shoppingCart = cloneDeep(this.state.shoppingCart)
    shoppingCart[id]--
    if (shoppingCart[id] === 0) {
      delete (shoppingCart[id])
    }
    this.setState({
      shoppingCart
    }, () => {
      this.handleTotalBill()
    })
  }

  handleTotalBill() {
    let { amount = 0 } = this.state
    let tempShoppingCart = cloneDeep(this.state.shoppingCart)
    let totalAmount = 0
    // Discount based on unique books count
    let getDiscount = function (count) {
      switch (count) {
        case 2:
          return 5;
        case 3:
          return 10;
        case 4:
          return 20;
        case 5:
          return 25;
        default:
          break;
      }
    }
    /* 
      Create sets of unique books to maximum discount calculation. 
      For example if cart has AABBCDE,
      first calculate price for ABCDE - 25% discount
      then price for AB - 5% discount
    */
    while (Object.keys(tempShoppingCart).length > 0) {
      let differentBooksCount = 0

      for (let book in tempShoppingCart) {
        differentBooksCount++
        tempShoppingCart[book]--
        if (tempShoppingCart[book] === 0) {
          delete (tempShoppingCart[book])
        }
      }
      amount = this.baseFare * differentBooksCount
      if (differentBooksCount > 1) {
        amount -= (getDiscount(differentBooksCount) / 100 * amount)
      }
      totalAmount += amount
    }
    this.setState({
      amount: totalAmount
    })
  }

  render() {
    let { shoppingCart, amount } = this.state
    let { books } = this.props
    let totalBooksCount = sum(values(shoppingCart))
    let totalDiscount = this.baseFare * totalBooksCount - amount
    return (
      <div className="bookStore">
        <h1>Book Shop</h1>
        <div className="bookStore-container">
          <div className="bookStore-list">
            {
              // Map through books master data to display book item
              map(books, (book) => {
                return (
                  <div className="bookStore-list-item" key={book.id}>
                    <ListItem
                      book={book}
                      handleAddToCart={this.handleAddToCart}
                      handleRemoveFromCart={this.handleRemoveFromCart}
                      shoppingCartValue={shoppingCart[book.id]}
                      baseBookFare={this.baseFare}
                    />
                  </div>
                )
              })
            }
          </div>
          <div className="bookStore-shoppingCart">
            {
              Object.keys(shoppingCart).length > 0 ? <React.Fragment>
                <h3>SHOPPING CART</h3>
                {
                  map(shoppingCart, (count, bookId) => {
                    return (
                      <div key={bookId} className="bookStore-shoppingCart-item">
                        <div>{get(find(books, ['id', bookId]), 'title', '')}</div>
                        <div>{count}</div>
                      </div>
                    )
                  })
                }
                <div className="bookStore-totalBill">
                  {
                    // Show discount and total amount fields only if any discount is applied
                    totalDiscount > 0 && <React.Fragment>
                      <div className="bookStore-totalBill-item">
                        <div>Total Amount:</div>
                        <div>$ {(this.baseFare * totalBooksCount).toFixed(2)}</div>
                      </div>
                      <div className="bookStore-totalBill-item">
                        <div>Discount:</div>
                        <div>- $ {totalDiscount.toFixed(2)}</div>
                      </div>
                    </React.Fragment>
                  }
                  <div className="bookStore-totalBill-item finalAmount">
                    <div>Final Amount:</div>
                    <div>$ {amount.toFixed(2)}</div>
                  </div>
                </div>
              </React.Fragment> :
                // Displayed when cart is empty
                <div>Your cart is empty!</div>
            }
          </div>
        </div>
      </div>
    )
  }
}

BookStore.propTypes = {
  books: PropTypes.arrayOf(PropTypes.shape({})).isRequired
};

export default BookStore
