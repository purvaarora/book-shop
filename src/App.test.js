import React from 'react';
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import App from './App'

const books = [
  {
    id: 'book1',
    amount: 8,
    title: 'Harry Potter and the Sorcerers Stone',
    author: 'J.K. Rowling'
  },
  {
    id: 'book2',
    amount: 8,
    title: 'Harry Potter and the Chamber of Secrets',
    author: 'J.K. Rowling'
  },
  {
    id: 'book3',
    amount: 8,
    title: 'Harry Potter and the Prisoner of Azkaban',
    author: 'J.K. Rowling'
  },
  {
    id: 'book4',
    amount: 8,
    title: 'Harry Potter and the Goblet of Fire',
    author: 'J.K. Rowling'
  },
  {
    id: 'book5',
    amount: 8,
    title: 'Harry Potter and the Order of the Phoenix',
    author: 'J.K. Rowling'
  },
]

describe('App component tests', () => {
  it('matches snapshot', () => {
    const component = renderer.create(
      <App />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })

  describe('handleAddToCart adds a book to shopping cart', () => {
    const wrapper = shallow(<App />);
    const handleAddToCartSpy = jest.spyOn(wrapper.instance(), "handleAddToCart");
    // const handleTotalBillSpy = jest.spyOn(wrapper.instance(), "handleTotalBill");
    it("handleAddToCart is called and item is added to shopping cart", () => {
      wrapper.instance().handleAddToCart('book1');
      expect(handleAddToCartSpy).toHaveBeenCalledTimes(1);
      handleAddToCartSpy.mockReset();
      handleAddToCartSpy.mockRestore();
      // expect(wrapper.state()).toEqual({ 'book1': 1 })
      // expect(handleTotalBillSpy).toHaveBeenCalledTimes(1);
      // handleTotalBillSpy.mockReset();
      // handleTotalBillSpy.mockRestore();
    });
  })

  describe('handleRemoveFromCart removed a book from shopping cart', () => {
    const wrapper = shallow(<App />);
    const handleRemoveFromCartSpy = jest.spyOn(wrapper.instance(), "handleRemoveFromCart");
    // const handleTotalBillSpy = jest.spyOn(wrapper.instance(), "handleTotalBill");
    it("handleRemoveFromCart is called and item is removed from shopping cart", () => {
      wrapper.instance().handleRemoveFromCart('book1');
      expect(handleRemoveFromCartSpy).toHaveBeenCalledTimes(1);
      handleRemoveFromCartSpy.mockReset();
      handleRemoveFromCartSpy.mockRestore();
      // expect(wrapper.state()).toEqual({ 'book1': 1 })
      // expect(handleTotalBillSpy).toHaveBeenCalledTimes(1);
      // handleTotalBillSpy.mockReset();
      // handleTotalBillSpy.mockRestore();
    });
  })
})




