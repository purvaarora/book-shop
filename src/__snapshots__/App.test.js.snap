// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component tests matches snapshot 1`] = `
<div
  className="bookStore"
>
  <h1>
    Book Shop
  </h1>
  <div
    className="bookStore-container"
  >
    <div
      className="bookStore-list"
    >
      <div
        className="bookStore-list-item"
      >
        <img />
        <div
          className="bookStore-list-item-description"
        >
          <div>
            <p>
              Harry Potter and the Sorcerers Stone
            </p>
            <p
              className="subText"
            >
              By 
              J.K. Rowling
            </p>
            <p>
              $ 
              8.00
            </p>
          </div>
          <div
            className="bookStore-list-item-actionButtons"
          >
            <button
              className="bookStore-list-item-actionButtons-remove"
              disabled={true}
              id="book1-remove"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <line
                  x1="8"
                  x2="16"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
            <input
              defaultValue={0}
              readOnly={true}
              value={0}
            />
            <button
              className="bookStore-list-item-actionButtons-add"
              id="book1-add"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <line
                  x1="12"
                  x2="12"
                  y1="8"
                  y2="16"
                />
                <line
                  x1="8"
                  x2="16"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        className="bookStore-list-item"
      >
        <img />
        <div
          className="bookStore-list-item-description"
        >
          <div>
            <p>
              Harry Potter and the Chamber of Secrets
            </p>
            <p
              className="subText"
            >
              By 
              J.K. Rowling
            </p>
            <p>
              $ 
              8.00
            </p>
          </div>
          <div
            className="bookStore-list-item-actionButtons"
          >
            <button
              className="bookStore-list-item-actionButtons-remove"
              disabled={true}
              id="book2-remove"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <line
                  x1="8"
                  x2="16"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
            <input
              defaultValue={0}
              readOnly={true}
              value={0}
            />
            <button
              className="bookStore-list-item-actionButtons-add"
              id="book2-add"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <line
                  x1="12"
                  x2="12"
                  y1="8"
                  y2="16"
                />
                <line
                  x1="8"
                  x2="16"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        className="bookStore-list-item"
      >
        <img />
        <div
          className="bookStore-list-item-description"
        >
          <div>
            <p>
              Harry Potter and the Prisoner of Azkaban
            </p>
            <p
              className="subText"
            >
              By 
              J.K. Rowling
            </p>
            <p>
              $ 
              8.00
            </p>
          </div>
          <div
            className="bookStore-list-item-actionButtons"
          >
            <button
              className="bookStore-list-item-actionButtons-remove"
              disabled={true}
              id="book3-remove"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <line
                  x1="8"
                  x2="16"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
            <input
              defaultValue={0}
              readOnly={true}
              value={0}
            />
            <button
              className="bookStore-list-item-actionButtons-add"
              id="book3-add"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <line
                  x1="12"
                  x2="12"
                  y1="8"
                  y2="16"
                />
                <line
                  x1="8"
                  x2="16"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        className="bookStore-list-item"
      >
        <img />
        <div
          className="bookStore-list-item-description"
        >
          <div>
            <p>
              Harry Potter and the Goblet of Fire
            </p>
            <p
              className="subText"
            >
              By 
              J.K. Rowling
            </p>
            <p>
              $ 
              8.00
            </p>
          </div>
          <div
            className="bookStore-list-item-actionButtons"
          >
            <button
              className="bookStore-list-item-actionButtons-remove"
              disabled={true}
              id="book4-remove"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <line
                  x1="8"
                  x2="16"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
            <input
              defaultValue={0}
              readOnly={true}
              value={0}
            />
            <button
              className="bookStore-list-item-actionButtons-add"
              id="book4-add"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <line
                  x1="12"
                  x2="12"
                  y1="8"
                  y2="16"
                />
                <line
                  x1="8"
                  x2="16"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        className="bookStore-list-item"
      >
        <img />
        <div
          className="bookStore-list-item-description"
        >
          <div>
            <p>
              Harry Potter and the Order of the Phoenix
            </p>
            <p
              className="subText"
            >
              By 
              J.K. Rowling
            </p>
            <p>
              $ 
              8.00
            </p>
          </div>
          <div
            className="bookStore-list-item-actionButtons"
          >
            <button
              className="bookStore-list-item-actionButtons-remove"
              disabled={true}
              id="book5-remove"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <line
                  x1="8"
                  x2="16"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
            <input
              defaultValue={0}
              readOnly={true}
              value={0}
            />
            <button
              className="bookStore-list-item-actionButtons-add"
              id="book5-add"
              onClick={[Function]}
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <line
                  x1="12"
                  x2="12"
                  y1="8"
                  y2="16"
                />
                <line
                  x1="8"
                  x2="16"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="bookStore-shoppingCart"
    >
      <div>
        Your cart is empty!
      </div>
    </div>
  </div>
</div>
`;
